## Home Credit Default Risk Prediction

### Project Overview

This project aims to develop multiple machine learning models to predict the likelihood of loan default for Home Credit Group clients. As a proof-of-concept (POC) for our startup's risk evaluation service, we'll create a suite of models that can help banks make more informed decisions about loan approvals. Our goal is to maximize accuracy in identifying both creditworthy clients and potential defaults, providing a robust and flexible solution for our potential banking clients.

### Why is this Important?

Home Credit specializes in lending to individuals with limited or no credit history (often referred to as "unbanked" or "underbanked"). They utilize diverse data sources, including phone and transaction data, to assess loan repayment probability. By developing accurate default prediction models, we can:

- **Expand financial inclusion:** Enable deserving clients who might be rejected by traditional methods to access credit.
- **Optimize risk management:** Help financial institutions make more informed lending decisions, reducing their exposure to potential losses.
- **Demonstrate our startup's capabilities:** Showcase our ability to translate complex business requirements into effective machine learning solutions.

### Dataset Description

We're using the Home Credit Default Risk dataset, which includes comprehensive information about loan applications and borrowers' credit histories. The data is structured across several interconnected tables:

1. **Application Data (`application_{train|test}.csv`):** The primary table containing loan application details such as age, income, employment history, etc.
2. **Bureau Data (`bureau.csv` and `bureau_balance.csv`):** Information about the applicants' previous loans from other financial institutions.
3. **Previous Applications (`previous_application.csv`):** Historical data on the applicants' past loan applications with Home Credit.
4. **Home Credit Loan Details:**
   - **POS and Cash Loans Balance (`pos_cash_balance.csv`):** Monthly balance snapshots of previous point-of-sale and cash loans.
   - **Credit Card Balance (`credit_card_balance.csv`):** Monthly balance snapshots of previous credit cards.
5. **Installments Payments (`installments_payments.csv`):** Payment history for previous loans at Home Credit.

### Our Approach

We'll use a comprehensive, multi-stage approach to develop our predictive models:

1. **Initial Data Exploration:**

   - Conduct thorough Exploratory Data Analysis (EDA) on each table.
   - Identify key variables and their distributions.
   - Investigate relationships between features and the target variable (loan default).
   - Check for data quality issues, missing values, and anomalies.

2. **Feature Engineering and Data Preprocessing:**

   - Create new features based on domain knowledge and initial insights.
   - Handle missing data and outliers.
   - Perform appropriate encoding for categorical variables and scaling for numerical features.

3. **Statistical Inference:**

   - Define the target population and formulate multiple statistical hypotheses.
   - Construct confidence intervals and conduct appropriate statistical tests.
   - Analyze correlations and other relationships between variables.

4. **Model Development:**

   - Implement multiple machine learning algorithms (e.g., Logistic Regression, Random Forests, Gradient Boosting).
   - Utilize cross-validation techniques to ensure model robustness.
   - Perform hyperparameter tuning to optimize model performance.

5. **Model Evaluation and Selection:**

   - Assess models using appropriate performance metrics (e.g., AUC-ROC, precision, recall).
   - Analyze feature importance and model interpretability.
   - Select the best-performing models for deployment.

6. **Model Deployment:**

   - Deploy the top-performing models to Google Cloud Platform.
   - Ensure models are accessible via HTTP requests for easy integration.

7. **Documentation and Presentation:**
   - Clearly document all steps, assumptions, and results.
   - Prepare visualizations and explanations of our findings.
   - Develop recommendations for potential clients based on our insights.

### Additional Notes

- **Data Source:** The dataset is from the [Home Credit Default Risk Kaggle Competition](https://www.kaggle.com/competitions/home-credit-default-risk/data).
- **Geographic Scope:** Home Credit operates primarily in CIS and Southeast Asian countries, including Kazakhstan, Russia, Vietnam, China, Indonesia, and the Philippines.
- **Product Categories:** Home Credit offers various loan products:
  - Revolving loans (credit cards)
  - Consumer loans (point-of-sale or POS loans)
  - Cash loans
- **Ethical Considerations:** We'll pay close attention to potential biases in our models and strive for fair lending practices.
- **Iterative Process:** Our approach will be flexible, allowing for iterations and refinements based on insights gained throughout the project.

**Home Credit Dataset Schema:**

![Home Credit Dataset Schema](../images/data-scheme.png)

This project will demonstrate our ability to handle complex, real-world data and deliver valuable insights and predictive capabilities to the financial sector.

By the end of this project, we aim to provide Home Credit with a reliable and easy-to-use tool to predict loan default risk, allowing them to make better lending decisions and potentially extend credit to more people who can manage it responsibly.

In [15]:
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [16]:
import pandas as pd
import polars as pl
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
import xgboost as xgb
import lightgbm as lgb
import sweetviz as sv
from matplotlib.colors import LinearSegmentedColormap
import matplotlib.pyplot as plt
import matplotlib as mpl
import matplotlib.pyplot as plt
from matplotlib.colors import LinearSegmentedColormap, ListedColormap
from ydata_profiling.report.structure.variables import render_real


from retail_bank_risk.data_preprocessing_utils import (
    handle_missing_values,
    simple_imputation,
    flag_anomalies,
    reduce_memory_usage_pl,
)
from retail_bank_risk.basic_visualizations_utils import (
    plot_correlation_matrix,
    plot_combined_histograms,
    plot_categorical_features_by_target,
    plot_combined_bar_charts,
    plot_combined_boxplots,
    plot_single_bar_chart,
    plot_feature_importances,
)
from retail_bank_risk.model_utils import (
    evaluate_model,
    extract_feature_importances,
)
from retail_bank_risk.advanced_visualizations_utils import (
    plot_model_performance,
    plot_combined_confusion_matrices,
    plot_roc_curve,
    plot_precision_recall_curve,
    plot_confusion_matrix,
    plot_learning_curve,
)

import pandas as pd
from ydata_profiling import ProfileReport
from IPython.display import display

For reproducibility, we set a random seed.


In [3]:
np.random.seed(42)

We then load the CSV files using Polars.


In [4]:
application_train = pl.read_csv("../data/application_train.csv")
application_test = pl.read_csv("../data/application_test.csv")
previous_application = pl.read_csv("../data/previous_application.csv")
pos_cash_balance = pl.read_csv("../data/pos_cash_balance.csv")
installments_payments = pl.read_csv("../data/installments_payments.csv")
bureau = pl.read_csv("../data/bureau.csv")
bureau_balance = pl.read_csv("../data/bureau_balance.csv")
credit_card_balance = pl.read_csv("../data/credit_card_balance.csv")

We've loaded all the data tables, but our initial focus will be on the application_train DataFrame. 

This table contains the core details of each loan application, which might be sufficient for building our first models. 

If we need more information to improve accuracy, we'll then explore the other tables and incorporate relevant data from them.

Before proceeding, we will reduce the memory footprint of the dataset by converting the data types to lower-precision alternatives, which will lead to faster processing times.

In [21]:
application_train = reduce_memory_usage_pl(application_train)
application_test = reduce_memory_usage_pl(application_test)

Size before memory reduction: 289.61 MB
Initial data types: Counter({Float64: 65, Int64: 41, String: 16})
Size after memory reduction: 113.52 MB
Final data types: Counter({Float32: 65, Int8: 35, Categorical(ordering='physical'): 16, Int32: 4, Int16: 2})
Size before memory reduction: 45.51 MB
Initial data types: Counter({Float64: 65, Int64: 40, String: 16})
Size after memory reduction: 17.94 MB
Final data types: Counter({Float32: 65, Int8: 34, Categorical(ordering='physical'): 16, Int32: 4, Int16: 2})


Next up, we can check the initial first rows of the dataset.

In [22]:
application_train.head()

shape: (5, 122)
┌────────────┬────────┬────────────┬───────────┬───┬───────────┬───────────┬───────────┬───────────┐
│ SK_ID_CURR ┆ TARGET ┆ NAME_CONTR ┆ CODE_GEND ┆ … ┆ AMT_REQ_C ┆ AMT_REQ_C ┆ AMT_REQ_C ┆ AMT_REQ_C │
│ ---        ┆ ---    ┆ ACT_TYPE   ┆ ER        ┆   ┆ REDIT_BUR ┆ REDIT_BUR ┆ REDIT_BUR ┆ REDIT_BUR │
│ i32        ┆ i8     ┆ ---        ┆ ---       ┆   ┆ EAU_WEEK  ┆ EAU_MON   ┆ EAU_QRT   ┆ EAU_YEAR  │
│            ┆        ┆ cat        ┆ cat       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ ---       │
│            ┆        ┆            ┆           ┆   ┆ f32       ┆ f32       ┆ f32       ┆ f32       │
╞════════════╪════════╪════════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
│ 100002     ┆ 1      ┆ Cash loans ┆ M         ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 1.0       │
│ 100003     ┆ 0      ┆ Cash loans ┆ F         ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0       │
│ 100004     ┆ 0      ┆ Revolving  ┆ M         ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0       │
│            ┆        ┆ loans      ┆           ┆   ┆           ┆           ┆           ┆           │
│ 100006     ┆ 0      ┆ Cash loans ┆ F         ┆ … ┆ null      ┆ null      ┆ null      ┆ null      │
│ 100007     ┆ 0      ┆ Cash loans ┆ M         ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0       │
└────────────┴────────┴────────────┴───────────┴───┴───────────┴───────────┴───────────┴───────────┘

The dataset contains numerous features. Let's examine the number of rows and features in both the application_train and application_test DataFrames.

In [6]:
print("Number of Rows and Columns in Application Train:", application_train.shape)
print("Number of Rows and Columns in Application Test:", application_test.shape)

Number of Rows and Columns in Application Train: (307511, 122)
Number of Rows and Columns in Application Test: (48744, 121)


**application_train:** 307,511 rows, 122 columns

**application_test:** 48,744 rows, 121 columns (excluding the TARGET variable)

We have a large training dataset with over **300k loan applications**. This provides a solid foundation for building robust machine learning models. The test dataset is well-sized for evaluating model performance on unseen data.

Both datasets have many features (**over 120**). This allows us to capture complex relationships and build accurate predictive models. However, the high number of features means we need to be careful in our selection and engineering process. This will help us avoid overfitting and keep our models interpretable.

Therefore, we will **manually select features based on their descriptions in the application_train and application_test DataFrames.**

1. **TARGET** (for training set only): Our dependent variable, shows if a client had payment difficulties.

2. **AMT_INCOME_TOTAL:** Income is key in determining loan repayment ability.

3. **AMT_CREDIT:** The loan amount directly relates to default risk.

4. **AMT_ANNUITY:** Annual loan payment affects repayment ability.

5. **NAME_CONTRACT_TYPE:** Different loan types may have varying risk profiles.

6. **CODE_GENDER:** Gender might influence repayment behavior.

7. **FLAG_OWN_CAR and FLAG_OWN_REALTY**: Asset ownership can indicate financial stability and potential collateral.

8. **CNT_CHILDREN:** Number of dependents affects financial obligations.

9. **NAME_EDUCATION_TYPE:** Education level often relates to income stability and financial literacy.

10. **DAYS_BIRTH:** Age is important in credit risk assessment.

11. **DAYS_EMPLOYED:** Employment duration can show job stability.

12. **NAME_INCOME_TYPE:** Income source affects its stability and reliability.

13. **NAME_FAMILY_STATUS:** Family status can impact financial responsibilities and stability.

14. **REGION_POPULATION_RELATIVE:** Economic opportunities may vary by region population.

15. **DAYS_LAST_PHONE_CHANGE:** Recent changes might suggest instability.

16. **EXT_SOURCE_1, EXT_SOURCE_2, EXT_SOURCE_3:** These scores likely include credit scores or similar risk assessments.

17. **OCCUPATION_TYPE:** Different jobs have varying levels of security and income stability.

18. **AMT_GOODS_PRICE:** For consumer loans, this shows the value of financed goods.

19. **OBS_30_CNT_SOCIAL_CIRCLE and DEF_30_CNT_SOCIAL_CIRCLE:** Financial behavior of one's social circle can indicate personal financial behavior.

20. **AMT_REQ_CREDIT_BUREAU_YEAR:** Number of credit inquiries in the past year can show recent financial activity or distress.

These features cover various aspects of a loan applicant's profile, including financial capacity, demographics, assets, employment, credit history, and behavior. This selection aims to provide a comprehensive view for assessing credit risk.

In [7]:
print(application_train.schema)

Schema({'SK_ID_CURR': Int64, 'TARGET': Int64, 'NAME_CONTRACT_TYPE': String, 'CODE_GENDER': String, 'FLAG_OWN_CAR': String, 'FLAG_OWN_REALTY': String, 'CNT_CHILDREN': Int64, 'AMT_INCOME_TOTAL': Float64, 'AMT_CREDIT': Float64, 'AMT_ANNUITY': Float64, 'AMT_GOODS_PRICE': Float64, 'NAME_TYPE_SUITE': String, 'NAME_INCOME_TYPE': String, 'NAME_EDUCATION_TYPE': String, 'NAME_FAMILY_STATUS': String, 'NAME_HOUSING_TYPE': String, 'REGION_POPULATION_RELATIVE': Float64, 'DAYS_BIRTH': Int64, 'DAYS_EMPLOYED': Int64, 'DAYS_REGISTRATION': Float64, 'DAYS_ID_PUBLISH': Int64, 'OWN_CAR_AGE': Float64, 'FLAG_MOBIL': Int64, 'FLAG_EMP_PHONE': Int64, 'FLAG_WORK_PHONE': Int64, 'FLAG_CONT_MOBILE': Int64, 'FLAG_PHONE': Int64, 'FLAG_EMAIL': Int64, 'OCCUPATION_TYPE': String, 'CNT_FAM_MEMBERS': Float64, 'REGION_RATING_CLIENT': Int64, 'REGION_RATING_CLIENT_W_CITY': Int64, 'WEEKDAY_APPR_PROCESS_START': String, 'HOUR_APPR_PROCESS_START': Int64, 'REG_REGION_NOT_LIVE_REGION': Int64, 'REG_REGION_NOT_WORK_REGION': Int64, 'LI

In [8]:
print(application_train.describe())

shape: (9, 123)
┌───────────┬───────────┬──────────┬───────────┬───┬───────────┬───────────┬───────────┬───────────┐
│ statistic ┆ SK_ID_CUR ┆ TARGET   ┆ NAME_CONT ┆ … ┆ AMT_REQ_C ┆ AMT_REQ_C ┆ AMT_REQ_C ┆ AMT_REQ_C │
│ ---       ┆ R         ┆ ---      ┆ RACT_TYPE ┆   ┆ REDIT_BUR ┆ REDIT_BUR ┆ REDIT_BUR ┆ REDIT_BUR │
│ str       ┆ ---       ┆ f64      ┆ ---       ┆   ┆ EAU_WEEK  ┆ EAU_MON   ┆ EAU_QRT   ┆ EAU_YEAR  │
│           ┆ f64       ┆          ┆ str       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ ---       │
│           ┆           ┆          ┆           ┆   ┆ f64       ┆ f64       ┆ f64       ┆ f64       │
╞═══════════╪═══════════╪══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
│ count     ┆ 307511.0  ┆ 307511.0 ┆ 307511    ┆ … ┆ 265992.0  ┆ 265992.0  ┆ 265992.0  ┆ 265992.0  │
│ null_coun ┆ 0.0       ┆ 0.0      ┆ 0         ┆ … ┆ 41519.0   ┆ 41519.0   ┆ 41519.0   ┆ 41519.0   │
│ t         ┆           ┆          ┆           ┆   ┆           ┆           

Next up we can convert them to Pandas dataframe objects for AutoEDA tools.
